### PR TITLE
feat: add "Did You Know" document

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze Lets see what happens :)

**Comment by mlutze** [comment]
I see 2 successful checks!

I guess you could also break `helloworld` somehow if you want a negative test. But things worked as expected in testing on my fork, so I don't think that's totally necessary.

**Comment by magnus-madsen** [comment]
@JonathanStarup @mlutze Review for typos and new ideas.

I was thinking that we could tweet out one line per week.

File: docs/DIDYOUKNOW.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+- Flix supports channels and processes, including the powerful `select` expression.
+
+- Flix supports first-class Datalog constraints, i.e. Datalog program fragments are values.
+
+- Flix supports compile-time checked stratified negation.
+
+- Flix supports partial application, i.e. a function can be called with fewer
+  arguments that its declared number of formal parameters.
+
+- the Flix type and effect system is based on Hindley-Miler.
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:60]: Milner

File: docs/DIDYOUKNOW.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  The same core type system that powers OCaml, Standard ML, and Haskell.
+
+- the Flix type and effect system is sound, i.e. if a program type checks
+  then a type error cannot occur at run-time.
+
+- the Flix type and effect system supports complete type inference, i.e.
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:66]: Is there a caveat here about ambiguous signatures that require annotations?

File: docs/DIDYOUKNOW.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+- Flix -- by design -- disallows allow unused declarations. This prevents bit rot.
+
+- Flix -- by design -- does not support function overloading. Instead, functions are given meaningful names, e.g.
+  `Map.insert` and `Map.insertWithKey`.
+
+- Flix -- by design -- does not support variadic functions. We believe such overloading
+  is unprincipled and unnecessary.
+
+- Flix -- by design -- does not support labelled arguments. Records can be used as an alternative and works
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:85]: Should we mention the syntactic sugar?

File: docs/DIDYOUKNOW.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+- Flix has no implicit global state. Any global state must be passed around explicitly.
+
+- Flix is one language. There are no pragmas or compiler flags to enable or disable features. 
+
+- Flix supports type parameter elision. That is, in most cases, a polymorphic function does 
+  can be written without explicitly declaring its type parameters.
+
+- the Flix type and effect system can enforce purity.
+
+- Flix supports effect polymorphism? For example, the effect of `List.map`
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:20]: Question mark?

File: docs/DIDYOUKNOW.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  can be written without explicitly declaring its type parameters.
+
+- the Flix type and effect system can enforce purity.
+
+- Flix supports effect polymorphism? For example, the effect of `List.map` 
+  depends on the effect of its argument.
+
+- every declaration in Flix is private by default.
+
+- In Flix nothing in is executed before main. There is no global state nor any static initializers.
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:25]: Maybe format `main`

File: docs/DIDYOUKNOW.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+- the Flix type and effect system supports complete type inference, i.e.
+  if a program is typeable then the type inference in the Flix compiler
+  will find the typing without a single annotation.
+
+- The Flix "Tips and Tricks"-section https://doc.flix.dev/tipstricks/ contains many useful niceties.
+
+- Flix -- by design -- has no implicit coercions (but does provide several explicit coercion functions).
+
+- Flix -- by design -- disallows unused variables and shadowed variables. Research has shown that such
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:74]: Whenever I see "research shows xyz" I can't help thinking "citation needed"

File: docs/DIDYOUKNOW.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  spanning more than 30,000 lines of code.
+
+- the Flix Prelude (the collection of functions which are imported by default) is 
+  deliberately kept minimal and contains less than 20 functions.
+
+- most higher-order functions in the Flix standard library are effect polymorphic
+  (they can be called with pure or impure functions). The exceptions are functions
+  that define (or use) equality, ordering, etc.
+
+- Mutating data structure operations end with an exclamation mark. For example `Array.transform!`.
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:112]: This parses better for me as "Operations that mutate data structures..."

File: docs/DIDYOUKNOW.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+- the Flix standard library uses records to avoid confusion when functions take multiple arguments of
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:114]: This overlaps with an earlier one. I think it's fine to keep both though.

**Review by mlutze — COMMENTED** [review]
There's a bit of inconsistency with capitalization and punctuation. Are these all completing the "Did you know..." question or are they all standalone statements? I vote for the latter.

A lot of these I think could be accompanied by links to principles or other pages with more information for the curious.

File: docs/DIDYOUKNOW.md (hunk 2 above)
Comment (magnus-madsen) [reply docs/DIDYOUKNOW.md:66]: Yes, but then again, we also require top-level type signatures.

File: docs/DIDYOUKNOW.md (hunk 6 above)
Comment (magnus-madsen) [reply docs/DIDYOUKNOW.md:74]: image

File: docs/DIDYOUKNOW.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+- Flix has received grants from the Independent Research Fund Denmark, from
+  Amazon Research, and from the Concordium Foundation.
+
+- Flix has been discussed on Reddit, HackerNews, and LambdaTheUltimate.
+
+- more than 40 people have contributed to the Flix compiler.
+
+- more than 1,500 pull requests have been merged into the Flix compiler.
+
+- Flix used to support uniform function call syntax (UFCS) but we could not
```
Comment (mlutze) [inline docs/DIDYOUKNOW.md:219]: I think a whole "History" section would be quite interesting

Comment (magnus-madsen) [reply docs/DIDYOUKNOW.md:221]: Added

**Comment by magnus-madsen** [comment]
@ninaandrup @jaschdoc If you could read through the document and provide some feedback that would be great. (even if it just that "that makes sense")

**Comment by mlutze** [comment]
What about other higher-level history like "development of Flix started in 1742 and was initially designed for helping build stables"? (except not made-up)

**Comment by magnus-madsen** [comment]
> What about other higher-level history like "development of Flix started in 1742 and was initially designed for helping build stables"? (except not made-up)

But in a different document / PR. (I see you are already gunning for that HOPL submission... ... ...)

File: docs/DIDYOUKNOW.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+  polymorphic effects, and first-class Datalog constraints.
+
+- Flix has no global state. Any state must be passed around explicitly.
+
+- Flix is one language. There are no pragmas or compiler flags to enable or
+  disable features.
+
+- Flix supports type parameter elision. That is, polymorphic functions can be
+  written without explicitly introducing their type parameters. For example,
+  `def map(f: a -> b, l: List[a]): List[b]`.
```
Comment (ninaandrup) [inline docs/DIDYOUKNOW.md:18]: Nice example

File: docs/DIDYOUKNOW.md [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+- the Flix type and effect system can enforce that a function argument is pure.
```
Comment (ninaandrup) [inline docs/DIDYOUKNOW.md:20]: Should every line start with a capital letter?

**Review by ninaandrup — COMMENTED** [review]
I think it looks nice!

**Comment by jaschdoc** [comment]
Looks good, but as Nina pointed out there are lots of line that start with lower-case letters. There are some "number" facts in there like the compiler is about 80.000 lines of code--maybe you could add that the packaged jar is just short of 35 mb? Not sure if that's relevant.

**Pull request merged**